runtime/hsa-runtime/core/inc/amd_xdna_driver.h

```diff
@@ -44,9 +44,35 @@
 
 #include <memory>
 #include <unordered_map>
+#include <x86intrin.h>
 
 #include "core/inc/driver.h"
 #include "core/inc/memory_region.h"
+#include "core/driver/xdna/uapi/amdxdna_accel.h"
+
+// Flushes the cache lines assocaited with a BO. This
+// is used to sync a BO without going to the xdna driver.
+// This is from the XRT KMQ shim.
+inline void
+clflush_data(const void *base, size_t offset, size_t len)
+{
+  static long cacheline_size = 0;
+
+  if (!cacheline_size) {
+    long sz = sysconf(_SC_LEVEL1_DCACHE_LINESIZE);
+    if (sz <= 0)
+      return;
+    cacheline_size = sz;
+  }
+
+  const char *cur = (const char *)base;
+  cur += offset;
+  uintptr_t lastline = (uintptr_t)(cur + len - 1) | (cacheline_size - 1);
+  do {
+    _mm_clflush(cur);
```

Comment:
Is this necessary? Why not use the IOCTL?

Also, please make this a static member method of XdnaDriver and follow the style used throughout, i.e., FlushCpuCache() or something similar.

Reply:
I want to be consistent with the XRT shim when we can be, which is defaulting to flushing in user space. This could also be done via the same ioctl but we get the benefit of removing a syscall from the critical path.

Reply:
Good point on the name, will change to align the style.

Reply:
Ok. Then you may want to move this into the utils lib. It already includes x86 instrinsics and CPU cache flushing isn't a strict function of the XDNA driver.

Reply:
Agreed. Addressed in c393105f70cdd24f4c3e81f10fc5111c28c197df.